Bluetooth: Controller: Fix building ticker without remainder support

Tronil opened the pull request:
A few things were not flagged by CONFIG_BT_TICKER_REMAINDER_SUPPORT that should have been; This caused building without it to fail

Tronil committed:
Bluetooth: Controller: Fix building ticker without remainder support

A few things were not flagged by CONFIG_BT_TICKER_REMAINDER_SUPPORT
that should have been; This caused building without it to fail

Initialize remainder to 0 where used with ticker_next_slot_get_ext(),
since it may not get set if remainder support is off

Signed-off-by: Troels Nilsson <trnn@demant.com>

cvinayak commented:
Hmmm, I thought nrf54L should have built without remainder support. Pleasedisable it for nRF54L

Tronil commented:
I found a problem with the expire_info with this change - fix coming

cvinayak commented:
Could you add a new commit with this below  change to enable bsim test to cover this:
```
diff --git a/subsys/bluetooth/controller/Kconfig.ll_sw_split b/subsys/bluetooth/controller/Kconfig.ll_sw_split
index 074549bcb3b..6ba68075d86 100644
--- a/subsys/bluetooth/controller/Kconfig.ll_sw_split
+++ b/subsys/bluetooth/controller/Kconfig.ll_sw_split
@@ -61,11 +61,12 @@ config BT_LLL_VENDOR_NORDIC
        select BT_CTLR_TIFS_HW_SUPPORT
        select BT_CTLR_ULL_LLL_PRIO_SUPPORT
 
-       select BT_TICKER_REMAINDER_SUPPORT
+       select BT_TICKER_REMAINDER_SUPPORT if !SOC_COMPATIBLE_NRF54LX
        select BT_TICKER_UPDATE if BT_BROADCASTER || BT_CONN || \
                                   (BT_OBSERVER && BT_CTLR_ADV_EXT)
        select BT_TICKER_START_REMAINDER if BT_CTLR_CENTRAL_ISO
-       select BT_TICKER_REMAINDER_GET if BT_BROADCASTER && BT_CTLR_ADV_EXT
+       select BT_TICKER_REMAINDER_GET if BT_TICKER_REMAINDER_SUPPORT && \
+                                         (BT_BROADCASTER && BT_CTLR_ADV_EXT)
        select BT_TICKER_LAZY_GET if BT_CTLR_ADV_PERIODIC || BT_CTLR_CENTRAL_ISO || BT_CTLR_SYNC_TRANSFER_SENDER
 
        select BT_TICKER_PREFER_START_BEFORE_STOP if BT_TICKER_SLOT_AGNOSTIC
diff --git a/tests/bsim/bluetooth/compile.nrf5340bsim_nrf5340_cpuapp.sh b/tests/bsim/bluetooth/compile.nrf5340bsim_nrf5340_cpuapp.sh
index 00dc5c4bdb6..96d4bd5db0e 100755
--- a/tests/bsim/bluetooth/compile.nrf5340bsim_nrf5340_cpuapp.sh
+++ b/tests/bsim/bluetooth/compile.nrf5340bsim_nrf5340_cpuapp.sh
@@ -16,6 +16,7 @@ app=tests/bsim/bluetooth/ll/conn conf_file=prj_split_privacy.conf sysbuild=1  co
 app=tests/bsim/bluetooth/ll/throughput sysbuild=1 compile
 app=tests/bsim/bluetooth/ll/multiple_id sysbuild=1 compile
 app=tests/bsim/bluetooth/ll/bis sysbuild=1 compile
+app=tests/bsim/bluetooth/ll/bis conf_overlay=overlay-ticker_expire_info.conf sysbuild=1 compile
 app=tests/bsim/bluetooth/ll/cis conf_overlay=overlay-acl_group.conf sysbuild=1 compile
 
 run_in_background ${ZEPHYR_BASE}/tests/bsim/bluetooth/samples/compile.sh
diff --git a/tests/bsim/bluetooth/compile.nrf54l15bsim_nrf54l15_cpuapp.sh b/tests/bsim/bluetooth/compile.nrf54l15bsim_nrf54l15_cpuapp.sh
index a17d9bf8677..43915d473b2 100755
--- a/tests/bsim/bluetooth/compile.nrf54l15bsim_nrf54l15_cpuapp.sh
+++ b/tests/bsim/bluetooth/compile.nrf54l15bsim_nrf54l15_cpuapp.sh
@@ -15,6 +15,8 @@ source ${ZEPHYR_BASE}/tests/bsim/compile.source
 
 app=tests/bsim/bluetooth/ll/throughput compile
 app=tests/bsim/bluetooth/ll/multiple_id compile
+app=tests/bsim/bluetooth/ll/bis compile
+app=tests/bsim/bluetooth/ll/bis conf_overlay=overlay-ticker_expire_info.conf compile
 
 run_in_background ${ZEPHYR_BASE}/tests/bsim/bluetooth/samples/compile.sh
 run_in_background ${ZEPHYR_BASE}/tests/bsim/bluetooth/audio/compile.sh
diff --git a/tests/bsim/bluetooth/tests.nrf5340bsim_nrf5340_cpuapp.txt b/tests/bsim/bluetooth/tests.nrf5340bsim_nrf5340_cpuapp.txt
index 539015d1428..80db7738abb 100644
--- a/tests/bsim/bluetooth/tests.nrf5340bsim_nrf5340_cpuapp.txt
+++ b/tests/bsim/bluetooth/tests.nrf5340bsim_nrf5340_cpuapp.txt
@@ -4,6 +4,7 @@ tests/bsim/bluetooth/ll/conn/tests_scripts/basic_conn_encrypted_split_privacy.sh
 tests/bsim/bluetooth/ll/throughput/
 tests/bsim/bluetooth/ll/multiple_id/
 tests/bsim/bluetooth/ll/bis/tests_scripts/broadcast_iso.sh
+tests/bsim/bluetooth/ll/bis/tests_scripts/broadcast_iso_ticker_expire_info.sh
 tests/bsim/bluetooth/samples/central_hr_peripheral_hr/
 tests/bsim/bluetooth/audio_samples/
 tests/bsim/bluetooth/audio/
diff --git a/tests/bsim/bluetooth/tests.nrf54l15bsim_nrf54l15_cpuapp.txt b/tests/bsim/bluetooth/tests.nrf54l15bsim_nrf54l15_cpuapp.txt
index 1df18c98a43..f1207aee21f 100644
--- a/tests/bsim/bluetooth/tests.nrf54l15bsim_nrf54l15_cpuapp.txt
+++ b/tests/bsim/bluetooth/tests.nrf54l15bsim_nrf54l15_cpuapp.txt
@@ -2,6 +2,8 @@
 # This file is used in CI to select which tests are run
 tests/bsim/bluetooth/ll/throughput/
 tests/bsim/bluetooth/ll/multiple_id/
+tests/bsim/bluetooth/ll/bis/tests_scripts/broadcast_iso.sh
+tests/bsim/bluetooth/ll/bis/tests_scripts/broadcast_iso_ticker_expire_info.sh
 tests/bsim/bluetooth/samples/central_hr_peripheral_hr/
 tests/bsim/bluetooth/audio_samples/
 tests/bsim/bluetooth/audio/
```

Tronil committed:
Bluetooth: Controller: Add tests covering ticker without remainder support

Update Kconfig to not select BT_TICKER_REMAINDER_SUPPORT for nrf54L

Add BSim tests to cover using a ticker without BT_TICKER_REMAINDER_SUPPORT
and expire info enabled

Signed-off-by: Troels Nilsson <trnn@demant.com>

Tronil commented:
> Could you add a new commit with this below change to enable bsim test to cover this:
> ...

Applied the patch locally, but it doesn't build correctly. I'll dig into it a bit to see if I can easily fix it.

cvinayak commented:
> > Could you add a new commit with this below change to enable bsim test to cover this:
> > ...
> 
> Applied the patch locally, but it doesn't build correctly. I'll dig into it a bit to see if I can easily fix it.

I had compile error as I prepared the diff without your PR, I can give a try using your branch tomorrow.

Tronil commented:
Got it building with a few more fixes; But the added tests are still failing for me unfortunately. I'll see if I can find out why tomorrow.

Tronil commented:
@cvinayak the two test cases added for nrf54l (broadcast_iso.sh and broadcast_iso_ticker_expire_info.sh) both fail without this change as well; Maybe we should leave those out for now so we can get this change in?

cvinayak commented:
Please use this fix:

```
diff --git a/tests/bsim/bluetooth/ll/bis/src/test_bis.c b/tests/bsim/bluetooth/ll/bis/src/test_bis.c
index 3b2b1a001c2..205fc2e8c71 100644
--- a/tests/bsim/bluetooth/ll/bis/src/test_bis.c
+++ b/tests/bsim/bluetooth/ll/bis/src/test_bis.c
@@ -193,7 +193,17 @@ static void iso_send(struct k_work *work)
                return;
        }
 
-       k_work_schedule(&iso_send_work, K_USEC(9970));
+       /* Ideally we can wait 10 ms to enqueue next SDU, but to mitigate clock
+        * drift between OS clock and Bluetooth Sleep Clock, lets send/enqueue
+        * early towards the Controller so that the SDU is not considered stale
+        * by the Controller.
+        *
+        * As we use `net_buf_alloc()` which will indirectly wait for Number of
+        * Completed Packet event processing in the Host leading to availability
+        * of next buffer, sending early will eventually get sync-ed with the
+        * interval at which the Controller is sending the SDU on-air.
+        */
+       k_work_schedule(&iso_send_work, K_USEC(9900));
 }
 #endif /* !CONFIG_TEST_LL_INTERFACE */
 
```

fabiobaltieri commented:
@cvinayak I moved the milestones to 4.2 since we are in RC3 now, feel free to comment if you think this should be considered a release blocker and go in instead, or just wait for 4.2 and eventually be backported.

cvinayak commented:
@fabiobaltieri  I am ok with v4.2.0 milestone, this is fixing an out-of-tree use I think.

@Tronil if you need it, please let us know :-)
